oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java

```diff
@@ -0,0 +1,197 @@
+             */
+            if (activeProductVersion == -1) {
+                indexVersionOperationList.add(new IndexVersionOperation(indexNameObjectList.get(i)));
+                if (indexNode.hasChildNode(IndexDefinition.STATUS_NODE)) {
+                    if (indexNode.getChildNode(IndexDefinition.STATUS_NODE)
+                            .getProperty(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP) != null) {
+                        String reindexCompletionTime = indexDefParentNode
+                                .getChildNode(PathUtils.getName(indexNameObjectList.get(i).getNodeName()))
+                                .getChildNode(IndexDefinition.STATUS_NODE)
+                                .getProperty(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP).getValue(Type.DATE);
+                        long reindexCompletionTimeInMillis = PurgeOldVersionUtils.getMillisFromString(reindexCompletionTime);
+                        long currentTimeInMillis = System.currentTimeMillis();
+                        if (currentTimeInMillis - reindexCompletionTimeInMillis > thresholdInMillis) {
+                            activeProductVersion = indexNameObjectList.get(i).getProductVersion();
+                        }
+                    } else {
+                        LOG.warn(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP
+                                + " property is not set for index " + indexNameObjectList.get(i).getNodeName());
+                    }
+                }
+            } else {
+            /*
+            Once we found active index, we mark all its custom version with DELETE
+            and OOTB same product version index with DELETE_HIDDEN_AND_DISABLE
+             */
+                if (indexNameObjectList.get(i).getProductVersion() == activeProductVersion
+                        && indexNameObjectList.get(i).getCustomerVersion() == 0) {
+                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObjectList.get(i));
+                    indexVersionOperation.setOperation(Operation.DELETE_HIDDEN_AND_DISABLE);
+                    indexVersionOperationList.add(indexVersionOperation);
+                } else {
+                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObjectList.get(i));
+                    indexVersionOperation.setOperation(Operation.DELETE);
+                    indexVersionOperationList.add(indexVersionOperation);
+                }
+            }
+        }
+        if (!isValidIndexVersionOperationList(indexVersionOperationList)) {
+            indexVersionOperationList = Collections.emptyList();
+        }
+        return indexVersionOperationList;
+    }
+
+    /*
+        returns IndexNameObjects in descending order of version i.e from latest to oldest
+    */
+    private static List<IndexName> getSortedVersionedIndexNameObjects(List<IndexName> indexNameObjectList) {
```

Comment:
I would rename to "sortedVersionedIndexNameObjects" because the method is modifying the list that was passed. (Which is fine). The method name "getSortedVersionedIndexNameObjects" to me means that it returns a different (cloned) list, and doesn't modify the list that was passed.